teachertool/src/components/MakecodeFrame.tsx

```diff
@@ -1,21 +1,28 @@
 /// <reference path="../../../built/pxteditor.d.ts" />
-import * as React from "react";
+import { useContext } from "react";
 import { setEditorRef } from "../services/makecodeEditorService";
-interface MakeCodeFrameProps {
-    pageSourceUrl: string;
-    tutorialEventHandler?: (event: pxt.editor.EditorMessageTutorialEventRequest) => void;
-}
+import { AppStateContext } from "../state/appStateContext";
+import { createIFrameUrl } from "../utils";
 
-export const MakeCodeFrame: React.FC<MakeCodeFrameProps> =
-    ( { pageSourceUrl} ) => {
+interface MakeCodeFrameProps {}
+export const MakeCodeFrame: React.FC<MakeCodeFrameProps> = () => {
+    const { state: teacherTool } = useContext(AppStateContext)
 
     const handleIFrameRef = (el: HTMLIFrameElement | null) => {
         setEditorRef(el ?? undefined);
     }
 
     /* eslint-disable @microsoft/sdl/react-iframe-missing-sandbox */
-    return <div className="makecode-frame-outer" style={{ display: "block" }}>
-        <iframe className="makecode-frame" src={pageSourceUrl} title={"title"} ref={handleIFrameRef}></iframe>
-    </div>
+    return (
+        <div className="makecode-frame-outer" style={{ display: "block" }}>
+            {teacherTool.projectMetadata && (
+                <iframe
+                    className="makecode-frame"
+                    src={createIFrameUrl(teacherTool.projectMetadata.id)}
```

Comment:
I know you commented about this in your PR description, but I think this can be put into a transform to keep the `MakeCodeFrame` as clean as possible. You can have the transform access the app state and then just return the url, so instead of the util `createIFrameUrl` being called in source, it would the the transform getting called.

Reply:
I'm not sure it's worth creating a transform that just replaces one one-line call with another one-line call, simply redirecting into a util function. My understanding is that it's reasonable for components to read state and use it as needed for the UI components like this, whereas a transform is more useful for larger operations that actually affect state. (And in terms of code-shareability, calling into a transform ties the component to the teachertool in much the same way that this does.)

In this case, since we're really just reading state (not changing it) and calling a util function, I'm skeptical that a transform is the right approach...unless there's something I'm missing, of course.

Reply:
My inclination is the opposite of @srietkerk's here. I'd suggest moving `createIFrameUrl` into this file as a local function. I don't think it needs to be called from anywhere else.

Transforms are for code that affects the state of the app in some way.

Reply:
Oh, right. Thank you for that clarification @thsparks, @eanders-ms. 

I think I'd be more inclined, then, to do as you said in your comment and pass the `teachertool.projectMetadata.id` into the `MakeCodeFrame` as a prop. I think this is a clean alternative and still allows for the `MakeCodeFrame` to be decoupled from  app state. I feel like having the prop be named `pageSourceUrl` would still be fine, as well, but maybe there's a better name.

Reply:
> I think I'd be more inclined, then, to do as you said in your comment and pass the `teachertool.projectMetadata.id` into the `MakeCodeFrame` as a prop.

State is available to all components globally. This makes passing state as props largely unnecessary. Why does this component need this state passed as a prop?

Reply:
So we can keep the `MakeCodeFrame` disconnected from state; but I guess if it's only one place of attachment, and we want to move the frame to someplace more general in the future, refactoring this won't be difficult.

Reply:
FWIW, I think the code that would be more valuable to share/generalize in the future would be the makecodeEditorService rather than this component. I'm inclined to agree with Eric here. It makes sense to pull the function into this component, so no-one else should know or care about the iframe specifics.

Reply:
Agreed, but if you have the component generic enough you can just ship the component and the service as a packaged deal. That said, as I mentioned before, it's likely something that can easily be removed if we want to rethink it, so I'm fine with leaving it as is since I'm in the minority.

For what it's worth, I'm also in favor of having the `createIFrameUrl` in the component. I don't think what I'm talking about and what Eric's suggesting is mutually exclusive.

Reply:
From a reuse perspective, makecodeEditorService.ts is of more value than this UI component, and is already quite portable.